internal/controller/imagerepository_controller.go

```diff
@@ -877,3 +879,222 @@ func (r *ImageRepositoryReconciler) ImageRepositoryForSameUrlExists(ctx context.
+		log.Error(err, "failed to get application pull secret", "secretName", applicationPullSecretName)
+		return err
+	}
+
+	pullSecret := &corev1.Secret{}
+	if err := r.Client.Get(ctx, types.NamespacedName{Name: pullSecretName, Namespace: namespace}, pullSecret); err != nil {
+		log.Error(err, "failed to get individual pull secret", "secretName", pullSecretName)
+		return err
+	}
+	if pullSecret.Type != corev1.SecretTypeDockerConfigJson {
+		log.Info("Skipping secret due to unexpected type", "secretName", pullSecretName, "type", pullSecret.Type)
+		return nil
+	}
+
+	dockerConfigDataBytes, ok := pullSecret.Data[corev1.DockerConfigJsonKey]
+	if !ok {
+		log.Info("Missing .dockerconfigjson key", "secretName", pullSecretName)
+		return nil
+	}
+
+	var newAuths dockerConfigJson
+	if err := json.Unmarshal(dockerConfigDataBytes, &newAuths); err != nil {
+		log.Error(err, "failed to unmarshal .dockerconfigjson", "secretName", pullSecretName)
+		return err
+	}
+
+	var existingAuths dockerConfigJson
+	if err := json.Unmarshal(applicationPullSecret.Data[corev1.DockerConfigJsonKey], &existingAuths); err != nil {
+		log.Error(err, "failed to unmarshal existing .dockerconfigjson", "secretName", applicationPullSecretName)
+		return err
+	}
+
+	changed := false
+	if existingAuths.Auths == nil {
+		existingAuths.Auths = make(map[string]struct {
+			Auth string `json:"auth"`
+		})
+	}
+
+	// If there are multiple pullsecrets for the same registry,
+	// they are overriden and we keep only the latest
+	for reg, entry := range newAuths.Auths {
+		if existingAuths.Auths[reg].Auth != entry.Auth {
+			existingAuths.Auths[reg] = entry
+			changed = true
+		}
+	}
+
+	if !changed {
+		return nil
+	}
+
+	// Marshal and update
+	mergedData, err := json.Marshal(existingAuths)
+	if err != nil {
+		log.Error(err, "failed to marshal updated docker config json")
+		return err
+	}
+	applicationPullSecret.Data[corev1.DockerConfigJsonKey] = mergedData
+
+	if err := r.Client.Update(ctx, applicationPullSecret); err != nil {
+		log.Error(err, "failed to update application pull secret", "secretName", applicationPullSecretName)
+		return err
+	}
+
+	log.Info("Application pull secret updated with new registry auth", "application", applicationName)
+	return nil
+
+}
+
+func (r *ImageRepositoryReconciler) removePullSecretFromApplicationPullSecret(ctx context.Context, applicationName string, namespace string, pullSecretName string, registryUrl string) error {
+	log := ctrllog.FromContext(ctx)
+
+	applicationPullSecretName := getApplicationPullSecretName(applicationName)
+	applicationPullSecret := &corev1.Secret{}
+	if err := r.Client.Get(ctx, types.NamespacedName{Name: applicationPullSecretName, Namespace: namespace}, applicationPullSecret); err != nil {
+		log.Error(err, "failed to get application pull secret", "secretName", applicationPullSecretName)
+		return err
```

Comment:
if application secret doesn't exist 404, we have nothing to remove so we can just return nil

if it fails on anything else then notfound return err

Reply:
as this would basically prevent IR to be removed, while not removing finalizer

Reply:
updated